Documentation/design-docs/native-hosting.md

```diff
@@ -0,0 +1,354 @@
+The library would be a dynamically loaded library (`.dll`, `.so`, `.dylib`). For ease of use there would be a header file for C/C++ apps as well as `.lib`/`.a` for easy linking.
+Native host would ship this library as part of the app. Unlike the `apphost`, `comhost` and `ijwhost`, the `nethost` will not be directly supported by the .NET Core SDK since it's target usage is not from .NET Core apps.
+
+The exact delivery mechanism is TBD (pending investigation), but it probably should include NuGet (for C++ projects) and plain `.zip` (for any consumer). The binary itself should be signed by Microsoft as there will be no support for modifying the binary as part of custom application build (unlike `apphost`).
+
+
+### Locate `hostfxr`
+``` C++
+int nethost_get_hostfxr_path(
+        char_t * result_buffer,
+        size_t buffer_size,
+        size_t * out_buffer_required_size,
+        const_t char * assembly_path);
+```
+
+This API locates the `hostfxr` and returns its path by calling the `result` function. (This callback design is chosen so that it's clear and easy to define memory ownership.)
+
+* `result_buffer` - Buffer that will be populated with the hostfxr path, including a null terminator.
+* `buffer_size` - Size of result_buffer in char_t units
+* `out_buffer_required_size` - Minimum required size in char_t units for a buffer to hold the hostfxr path
+* `assembly_path` - Optional. Path to the component's assembly. Whether or not this is specified determines the behavior for locating the hostfxr library.
+  * If `nullptr`, `hostfxr` is located using the environment variable or global registration
+  * If specified, `hostfxr` is located as if the `assembly_path` is and application with `apphost`
+
+`nethost` library uses the `__stdcall` calling convention.
+
+
+## Improve API to run application and load components
+
+### Goals
+* All hosts should be able to use the new API (whether they will is a separate question as the old API has to be kept for backward compat reasons)
+* Hide implementation details as much as possible
+  * Make the API generally easier to use/understand
+  * Give the implementation more freedom
+  * Allow future improvements without breaking the API
+  * Consider explicitly documenting types of behaviors which nobody should take dependency on (specifically failure scenarios)
+* Extensible
+
+### New scenarios
+The API should allow these scenarios:
+* Runtime properties
+  * Specify additional runtime properties from the native host
+  * Implement conflict resolution for runtime properties
+  * Inspect calculated runtime properties (the ones calculated by `hostfxr`/`hostpolicy`)
+
+It should be possible to ship with only some of these supported, then enable more scenarios later on.
+
+All the proposed APIs will be exports of the `hostfxr` library and will use the same calling convention as existing `hostfxr` exports. The names shown are the exact export names (no mangling).
```

Comment:
>exact export names (no mangling).

This will require some additional linker flags for x86 if `__stdcall` is defined. It is largely free for `__cdecl`.

Reply:
I left that out - the intent is to use exactly the same export mechanism as the other `hostfxr` exports.